nova/virt/vmwareapi/vmops.py

```diff
@@ -3306,3 +3408,63 @@ def _sync_sync_server_group(context, sg_uuid):
             LOG.debug('Sync for server-group %s done', sg_uuid)
 
         _sync_sync_server_group(context, sg_uuid)
+
+    def place_vm(self, context, instance):
+        # We currently only fill the bare-minimum to get a placement.
+        # The datastore for the VM is selected as on instance creation,
+        # instead of allowing placevm to decide it (which may be better)
+        # We also do not pass the information about the NICs and other
+        # attached disks, which may give the placement a better information
+        client_factory = self._session.vim.client.factory
+        flavor = instance.flavor
+        image_meta = instance.image_meta
+        image_info = images.VMwareImage.from_image(context,
+                                                instance.image_ref,
+                                                image_meta)
+
+        extra_specs = self._get_extra_specs(flavor, image_meta)
+
+        vi = self._get_vm_config_info(instance, image_info, extra_specs)
+
+        vm_folder = self._get_project_folder(vi.dc_info,
+            project_id=instance.project_id, type_='Instances')
+
+        relocate_spec = vm_util.relocate_vm_spec(
+            client_factory,
+            res_pool=self._root_resource_pool,
+            datastore=vi.datastore.ref,
+            disk_move_type="moveAllDiskBackingsAndDisallowSharing",
+            folder=vm_folder)
+
+        placement_spec = client_factory.create("ns0:PlacementSpec")
+        placement_spec.placementType = "relocate"
+        # Maybe we want to allow that? Default is True
+        # placement_spec.disallowPrerequisiteMoves = False
+        placement_spec.relocateSpec = relocate_spec
+        # So we do not place the vm on a failover host
+        placement_spec.hosts, _ = \
+            vm_util.get_hosts_and_reservations_for_cluster(
+                self._session, self._cluster)
+
+        # Sets cpuAllocation, memoryAllocation, numCPUs, memoryMB
+        config_spec = vm_util.get_vm_resize_spec(client_factory,
+                                                 int(flavor.vcpus),
+                                                 int(flavor.memory_mb),
+                                                 extra_specs)
+        # The last mandatory field for config_spec per doc
+        config_spec.version = extra_specs.hw_version
+        placement_spec.configSpec = config_spec
+
+        vm_group_name = self._get_admin_group_name_for_instance(instance)
+        if vm_group_name:
+            placement_rules = []
+            placement_spec.rules = placement_rules
+            cluster_rules = cluster_util.fetch_cluster_rules(self._session,
+                                                             self._cluster)
+            for rule in six.itervalues(cluster_rules):
+                if getattr(rule, "vmGroupName", None) == vm_group_name:
+                    placement_rules.append(rule)
+
```

Comment:
> You currently look for the rule containing the DRS group. [...]

Sure, that is work which needs to be done, but I am not convinced that this patch has to take care of that.
Just from that description, it even sounds that it is mostly compatible, save for some locking, which might needs to get introduced.

Reply:
Fair point. I just thought that code looked unnecessarily complicated, as it would just have to call `cluster_util.get_rule()`. Then let's merge and update later, when we have the name.